File: typescript/packages/core/test/team/crash-kit.ts
import { z } from "zod";
import type { RunResult, TeamAgent } from "../../src";
import { hostRunner } from "../../src/agent/registry";
import { type Store, storeOf } from "../../src/agent/sqlite";
import {
  type BranchId,
  type TeamId,
  TeamId as TeamIdSchema,
  ThreadId,
} from "../../src/log";
import {
  LogStore,
  memoryArtifacts,
  type SqlValue,
  type StoreDriver,
  type Tx,
} from "../../src/store";
import { openBunSqlite } from "../../src/store/bun-sqlite";
import { unwrap } from "../store/helpers";
import { query } from "./kit";

// Shared by the team crash drills: a driver that dies once inside one commit point's transaction,
// and the restart, the host's recovery of the lead's open run with no new input.

export class Crash extends Error {}

export type Point = {
  readonly name: string;
  readonly at: (sql: string, params: readonly SqlValue[]) => boolean;
};

const utf8 = new TextDecoder();

/** A statement parameter holds `text` (a string, or JSON bytes). */
export const mentions = (params: readonly SqlValue[], text: string): boolean =>
  params.some(
    (p) =>
      (typeof p === "string" && p.includes(text)) ||
      (p instanceof Uint8Array && utf8.decode(p).includes(text)),
  );

/**
 * The same database, through a driver that dies once at `point`: that transaction rolls back and
 * the run rejects. Work still in flight may go on, as a sibling process's would.
 */
export function crashing(base: StoreDriver, point: Point): StoreDriver {
  let crashed = false;
  const wrap = (tx: Tx): Tx => ({
    ...tx,
    run: (sql, params = []) => {
      if (!crashed && point.at(sql, params)) {
        crashed = true;
        throw new Crash(`killed at ${point.name}`);
      }
      return tx.run(sql, params);
    },
    transaction: (fn) => tx.transaction((inner) => fn(wrap(inner))),
  });
  return {
    ...base,
    transaction: (fn, options) =>
      base.transaction((tx) => fn(wrap(tx)), options),
  };
}

const Row = z.object({ thread_id: ThreadId, team_id: z.string() });

/** A drill's database, its log store, and the host's restart of the lead. */
export type Drill = {
  readonly db: StoreDriver;
  readonly log: LogStore;
  /** The store over the drill's database, through `driver`. */
  readonly open: (driver: StoreDriver) => Promise<Store>;
  readonly restart: (lead: TeamAgent) => Promise<{
    readonly result: RunResult<unknown>;
    readonly branch: BranchId;
    readonly team: TeamId;
  }>;
};

/** One database, opened as a store through any driver over it. */
export async function drill(): Promise<Drill> {
  const db = openBunSqlite(":memory:");
  const artifacts = memoryArtifacts();
  const log = unwrap(await LogStore.open(db, Date.now, artifacts));
  const open = async (driver: StoreDriver): Promise<Store> =>
    storeOf({
      log: unwrap(await LogStore.open(driver, Date.now, artifacts)),
      artifacts,
    });
  /**
   * The host's recovery of the lead's open run, with no new input, once the dead process's leases
   * have expired.
   */
  const restart: Drill["restart"] = async (lead) => {
    await db.transaction((tx) =>
      tx.run("UPDATE leases SET expires_at = 0", []),
    );
    const [row] = z
      .array(Row)
      .parse(
        await query(
          db,
          "SELECT lead_thread_id AS thread_id, team_id FROM teams",
          [],
        ),
      );
    if (row === undefined)
      throw new Error("the lead's first append opened its team");
    const store = await open(db);
    const branch = unwrap(await log.mainBranch(row.thread_id));
    const runner = hostRunner(lead);
    if (runner === undefined)
      throw new Error("agent() registers a host runner");
    const result = await runner.execute(
      {
        store,
        principal: { issuer: "api", tenant: "local", subject: "operator" },
        thread: { id: row.thread_id, branch, store },
      },
      [],
    );
    return {
      result,
      branch,
      team: TeamIdSchema.parse(row.team_id),
    };
  };
  return { db, log, open, restart };
}

File: python/tests/team/crash_kit.py
"""Shared by the team crash drills: a connection that dies once inside one commit point's
transaction, and the restart, the host's recovery of the lead's open run with no new input (the
same as TypeScript's test/team/crash-kit.ts)."""

import sqlite3
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from functools import partial
from pathlib import Path

import pytest

from threads import Agent, Principal, RunResult, Store, sqlite
from threads.agents.run import RunOptions, execute
from threads.agents.store import open_store
from threads.log import BranchId, ThreadId
from threads.result import Ok
from threads.store import SqliteStore
from threads.store.sql import text_of
from threads.thread.handle import Thread

OPERATOR = Principal(issuer="api", tenant="local", subject="operator")


class CrashError(Exception):
    """The process dying at a commit point."""


type At = Callable[[sqlite3.Connection, str, Sequence[object]], bool]


@dataclass(frozen=True, slots=True)
class Point:
    name: str
    at: At


def mentions(params: Sequence[object], text: str) -> bool:
    """A statement parameter holds `text` (a string, or JSON bytes)."""
    return any(
        (isinstance(p, str) and text in p) or (isinstance(p, bytes) and text.encode() in p)
        for p in params
    )


class _Crashing(sqlite3.Connection):
    """A connection that dies once, at `point`."""

    point: At | None = None

    def execute(self, sql: str, parameters: Sequence[object] = (), /) -> sqlite3.Cursor:  # type: ignore[override] - narrowed for the drill
        point = type(self).point
        if point is not None and point(self, sql, parameters):
            type(self).point = None
            raise CrashError(sql)
        return super().execute(sql, parameters)


async def crashing(where: Path, monkeypatch: pytest.MonkeyPatch, point: Point) -> Store:
    """A store on `where` whose connection dies once at `point`."""
    store = sqlite(str(where))
    with monkeypatch.context() as m:
        m.setattr(sqlite3, "connect", partial(sqlite3.connect, factory=_Crashing))
        await open_store(store)
    _Crashing.point = point.at
    return store


def reached() -> bool:
    """Whether the drill reached its commit point."""
    return _Crashing.point is None


@dataclass(frozen=True, slots=True)
class Restarted:
    result: RunResult[str]
    sq: SqliteStore
    lead: BranchId
    team: str


async def restart(where: Path, lead: Agent[None, str]) -> Restarted:
    """The host's recovery of the lead's open run with no new input, on a fresh connection."""
    store = sqlite(str(where))
    sq = await open_store(store)
    teams = await sq.run(
        lambda c: c.execute("SELECT lead_thread_id, team_id FROM teams").fetchall()
    )
    ((lead_column, team_column),) = teams
    lead_thread, team = text_of(lead_column), text_of(team_column)
    branch = await sq.root(ThreadId(lead_thread))
    assert isinstance(branch, Ok)
    thread = Thread(ThreadId(lead_thread), branch.value, store)
    options: RunOptions[None] = {"store": store, "principal": OPERATOR, "thread": thread}
    result = await execute(lead.definition, None, options, None, lambda _e: None)
    return Restarted(result, sq, branch.value, team)
